src/action.ts

```diff
@@ -116,57 +143,180 @@ async function setupBinaryInEnv(binary: DownloadedBinary) {
+        throw new Error(
+          `Authentication failed when accessing '${owner}/${repo}'. Please check your token permissions.`
+        );
+      } else {
+        throw error;
+      }
+    }
   }
 
-  core.info(`Downloading ${owner}/${repo} version ${tag}`);
+  async validateAccess(owner: string, repo: string): Promise<void> {
+    try {
+      await this.octokit.rest.repos.get({
+        owner,
+        repo,
+      });
+    } catch (error: any) {
+      if (error.status === 404 || error.message.includes("Not Found")) {
+        if (owner !== "gruntwork-io") {
+          core.warning(
+            `Cannot validate access to '${owner}/${repo}' repository. This may be due to token permissions or repository visibility. Proceeding with download attempt.`
+          );
+          return;
+        }
+        throw Error(
+          `Cannot access the '${repo}' repository. This could indicate: 1) The repository doesn't exist, 2) Your token doesn't have access to this repository, or 3) Your token lacks the 'repo' scope for private repositories. Please check your token permissions and repository access. If you continue to have issues and are a Gruntwork customer, contact support@gruntwork.io.`
+        );
+      } else {
+        throw error;
+      }
+    }
+  }
+}
 
-  const getReleaseResponse = await octokit.rest.repos.getReleaseByTag({
-    owner,
-    repo,
-    tag,
-  });
+function createGitHubProvider(config: GitHubConfig): GitHubProviderInterface {
+  return new GitHubProvider(config);
+}
 
+function findCompatibleAsset(release: Release, owner: string, repo: string, tag: string): ReleaseAsset {
   const re = new RegExp(`${osPlatform()}.*${arch()}`);
-  const asset = getReleaseResponse.data.assets.find((obj: any) => re.test(obj.name));
+  const asset = release.assets.find((obj: ReleaseAsset) => re.test(obj.name));
 
   if (!asset) {
     throw new Error(`Can not find ${owner}/${repo} release for ${tag} in platform ${re}.`);
   }
 
-  // Use @actions/tool-cache to download the binary from GitHub
-  const downloadedPath = await toolCache.downloadTool(
-    asset.url,
-    // Don't set a destination path. It will default to a temporary one.
-    undefined,
-    `token ${token}`,
-    {
-      accept: "application/octet-stream",
-    }
-  );
+  return asset;
+}
 
-  core.debug(`${owner}/${repo}@'${tag}' has been downloaded at ${downloadedPath}`);
+function determineDownloadConfig(
+  asset: ReleaseAsset,
+  repo: string,
+  token: string
+): { assetUrl: string; authHeader: string | undefined; headers: Record<string, string> } {
+  // Always use the API URL for consistent authentication behavior
+  const assetUrl = asset.url;
+
+  core.debug(`Selected asset URL for ${assetUrl} (using API URL)`);
+
+  // Determine authentication and headers
+  let authHeader: string | undefined;
+  const headers: Record<string, string> = {};
+
+  // Asset API URL: always needs proper authentication
+  if (token) {
```

Comment:
I'm confused. Why are we trying to not set the header if token isn't specified? Shouldn't it always be specified?